components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/ClaimUtil.java

```diff
@@ -305,6 +316,56 @@ private static String getSpMappedRoleClaim(ServiceProvider serviceProvider,
         return OIDCClaimUtil.getServiceProviderMappedUserRoles(serviceProvider, rolesList, claimSeparator);
     }
 
+    private static List<String> filterSharedRoles(ServiceProvider serviceProvider,
+                                                  List<String> userRoles) throws FrameworkException {
+
+        List<String> rolesAssociatedWithApp = new ArrayList<>();
+        if (CollectionUtils.isNotEmpty(userRoles)) {
+            rolesAssociatedWithApp = new ArrayList<>(userRoles);
+            /*
+             Check whether the application is a fragment application and if so, add only the shared roles
+             to the user's role list.
+            */
+            if (serviceProvider.getSpProperties() != null && Arrays.stream(serviceProvider.getSpProperties()).anyMatch(
+                    spProperty -> ApplicationConstants.IS_FRAGMENT_APP.equals(spProperty.getName())
+                            && StringUtils.equals(Boolean.TRUE.toString(), spProperty.getValue()))) {
+                if (RoleConstants.ORGANIZATION.equalsIgnoreCase(serviceProvider.getAssociatedRolesConfig().
+                        getAllowedAudience())) {
+                    for (String roleName : rolesAssociatedWithApp) {
+                        try {
+                            String roleId = OAuth2ServiceComponentHolder.getInstance().getRoleManagementServiceV2().
+                                    getRoleIdByName(removeInternalDomain(roleName), RoleConstants.ORGANIZATION,
+                                            serviceProvider.getTenantDomain(), serviceProvider.getTenantDomain());
+                            if (roleId != null) {
+                                Role role = OAuth2ServiceComponentHolder.getInstance().getRoleManagementServiceV2().
+                                        getRole(roleId, serviceProvider.getTenantDomain());
+                                List<RoleProperty> roleProperties = role.getRoleProperties();
+                                boolean isSharedRole = roleProperties.stream().anyMatch(roleProperty ->
+                                        RoleConstants.IS_SHARED_ROLE_PROP_NAME.equals(roleProperty.getName()) &&
+                                                Boolean.TRUE.toString().equals(roleProperty.getValue()));
+                                if (!isSharedRole) {
+                                    rolesAssociatedWithApp.remove(roleName);
+                                }
+                            }
+                        } catch (IdentityRoleManagementException e) {
+                            throw new FrameworkException("Error while getting the role details of role : " +
+                                    roleName, e);
+                        }
+                    }
+                }
+            }
+        }
+        return rolesAssociatedWithApp;
```

Comment:
Isn't that non shared org audience role consuming apps get empty list of roles due to this?